File: packages/go/internal/seguranca/ataques/ataque_rotacional.go
package ataques

import (
	"bytes"
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesRotacional configura o AtaqueRotacional.
type OpcoesRotacional struct {
	Tamanho int
}

// AtaqueRotacional testa se a cifra é simétrica a deslocamentos circulares de
// key/iv: keystream(rot(key), rot(iv)) seria uma rotação de keystream(key, iv).
type AtaqueRotacional struct {
	tamanho int
}

// NovoAtaqueRotacional cria o ataque. Sem argumento, usa tamanho = 64.
func NovoAtaqueRotacional(opcoes ...OpcoesRotacional) *AtaqueRotacional {
	o := OpcoesRotacional{Tamanho: 64}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueRotacional{tamanho: o.Tamanho}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueRotacional) Nome() string {
	return "Rotacional/slide (simetria por rotação)"
}

// rotacionar desloca circularmente um buffer k posições.
func rotacionar(buf []byte, k int) []byte {
	n := len(buf)
	out := make([]byte, n)
	for i := 0; i < n; i++ {
		out[i] = buf[(i+k)%n]
	}
	return out
}

// Executar roda o ataque.
func (a *AtaqueRotacional) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())
	iv := seguranca.RandomBytes(alvo.TamanhoIv())
	ks1 := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanho)

	coincidencias := []string{}
	for k := 1; k < len(chave); k++ {
		ks2 := alvo.GerarKeystreamBruto(rotacionar(chave, k), rotacionar(iv, k), "enc", a.tamanho)

		if bytes.Equal(ks2, ks1) {
			coincidencias = append(coincidencias, fmt.Sprintf("rotação %d: keystream idêntico", k))
			continue
		}
		alvoRot := rotacionar(ks1[:32], k)
		if bytes.Equal(ks2[:32], alvoRot) {
			coincidencias = append(coincidencias, fmt.Sprintf("rotação %d: keystream rotacionado", k))
		}
	}

	vulneravel := len(coincidencias) > 0
	severidade := seguranca.SevInfo
	detalhes := fmt.Sprintf("Nenhuma das %d rotações de byte reproduziu o keystream", len(chave)-1)
	if vulneravel {
		severidade = seguranca.SevCritica
		detalhes = "Simetria rotacional encontrada: " + strings.Join(coincidencias[:min(5, len(coincidencias))], "; ")
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes:   detalhes,
		Dados:      map[string]any{"coincidencias": coincidencias},
	}, nil
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueRotacional.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Ataque rotacional/slide: se a cifra for simétrica a deslocamentos circulares
 * de key/iv, então `keystream(rot(key), rot(iv))` seria uma rotação de
 * `keystream(key, iv)` - uma estrutura explorável. Testa todas as rotações de
 * byte e também coincidência direta do keystream.
 */
public class AtaqueRotacional implements AtaqueInterface {
    private final int tamanho;

    public AtaqueRotacional() {
        this(64);
    }

    public AtaqueRotacional(int tamanho) {
        this.tamanho = tamanho;
    }

    @Override
    public String nome() {
        return "Rotacional/slide (simetria por rotação)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] iv = Util.randomBytes(alvo.tamanhoIv());
        byte[] ks1 = alvo.gerarKeystreamBruto(chave, iv, "enc", this.tamanho);

        List<String> coincidencias = new ArrayList<>();
        for (int k = 1; k < chave.length; k++) {
            byte[] ks2 = alvo.gerarKeystreamBruto(
                    this.rotacionar(chave, k), this.rotacionar(iv, k), "enc", this.tamanho);

            if (Arrays.equals(ks2, ks1)) {
                coincidencias.add("rotação " + k + ": keystream idêntico");
                continue;
            }
            // metade inicial de ks1 rotacionada deve bater com ks2 se houver simetria
            byte[] alvoRot = this.rotacionar(Arrays.copyOfRange(ks1, 0, 32), k);
            if (Arrays.equals(Arrays.copyOfRange(ks2, 0, 32), alvoRot)) {
                coincidencias.add("rotação " + k + ": keystream rotacionado");
            }
        }

        boolean vulneravel = !coincidencias.isEmpty();

        Map<String, Object> dados = new HashMap<>();
        dados.put("coincidencias", coincidencias);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.CRITICA : Severidade.INFO,
                vulneravel
                        ? "Simetria rotacional encontrada: "
                                + String.join("; ", coincidencias.subList(0, Math.min(5, coincidencias.size())))
                        : "Nenhuma das " + (chave.length - 1)
                                + " rotações de byte reproduziu o keystream",
                dados);
    }

    private byte[] rotacionar(byte[] buf, int k) {
        int n = buf.length;
        byte[] out = new byte[n];
        for (int i = 0; i < n; i++) {
            out[i] = buf[(i + k) % n];
        }
        return out;
    }
}
